numpy/_core/src/multiarray/datetime.c

```diff
@@ -2818,85 +2819,247 @@ convert_datetime_to_pyobject(npy_datetime dt, PyArray_DatetimeMetaData *meta)
 }
 
 /*
- * Converts a timedelta into a PyObject *.
+ * We require that if d is a PyDateTime, then
+ * hash(numpy.datetime64(d)) == hash(d).
+ * Where possible, convert dt to a PyDateTime and hash it.
  *
- * Not-a-time is returned as the string "NaT".
- * For microseconds or coarser, returns a datetime.timedelta.
- * For units finer than microseconds, returns an integer.
+ * NOTE: "equals" across PyDate, PyDateTime and np.datetime64 is not transitive:
+ * datetime.datetime(1970, 1, 1) == np.datetime64(0, 'us')
+ * np.datetime64(0, 'us') == np.datetime64(0, 'D')
+ * datetime.datetime(1970, 1, 1) != np.datetime64(0, 'D')  # date, not datetime!
+ *
+ * But:
+ * datetime.date(1970, 1, 1) == np.datetime64(0, 'D')
+ *
+ * For hash(datetime64(0, 'D')) we could return either PyDate.hash or PyDateTime.hash.
+ * We choose PyDateTime.hash to match datetime64(0, 'us')
  */
-NPY_NO_EXPORT PyObject *
-convert_timedelta_to_pyobject(npy_timedelta td, PyArray_DatetimeMetaData *meta)
+NPY_NO_EXPORT npy_hash_t
+datetime_hash(PyArray_DatetimeMetaData *meta, npy_datetime dt)
 {
-    npy_timedelta value;
-    int days = 0, seconds = 0, useconds = 0;
+    npy_hash_t res, obj_hash;
+    npy_datetimestruct dts;
 
-    /*
-     * Convert NaT (not-a-time) into None.
-     */
-    if (td == NPY_DATETIME_NAT) {
-        Py_RETURN_NONE;
+    if (dt == NPY_DATETIME_NAT || meta->base == NPY_FR_GENERIC) {
+        return -2;
     }
 
-    /*
-     * If the type's precision is greater than microseconds, is
-     * Y/M/B (nonlinear units), or is generic units, return an int
-     */
-    if (meta->base > NPY_FR_us ||
-                    meta->base == NPY_FR_Y ||
-                    meta->base == NPY_FR_M ||
-                    meta->base == NPY_FR_GENERIC) {
-        return PyLong_FromLongLong(td);
+    if (NpyDatetime_ConvertDatetime64ToDatetimeStruct(meta, dt, &dts) < 0) {
+        return -1;
     }
 
-    value = td;
+    res = (dts.ps * HASH_MULT + dts.as) * HASH_MULT;
+
+    if (dts.year < 1 || dts.year > 9999) {
+        res = ((((((res + dts.year)
+                   * HASH_MULT + dts.month)
+                  * HASH_MULT + dts.day)
+                 * HASH_MULT + dts.hour)
+                * HASH_MULT + dts.min)
+               * HASH_MULT + dts.sec)
+            * HASH_MULT + dts.us;
```

Comment:
FWIW, I would much prefer to use the tuple update pattern as here: https://github.com/numpy/numpy/pull/24408/files#diff-c32b1134e6b616b5aca0e41218075be0ae2cb509a03e46b1716e1988a93fafd8R643-R656

I don't trust `HAS_MULT` to do a half-decent job, TBH.  The code also doesn't take into account that that the size of the hash could be 32bit.

Reply:
Haha, yep, actually in an earlier version of this patch I was converting the datetimestruct into a PyBytes, and using python's own bytes_hash on it. (see [here](https://github.com/numpy/numpy/commit/bf3b9957b27f17f7dde096cd4ea7c9bdf2d1b3b4) ).

The version using PyBytes has less code than the "tuple update pattern". I'll resurrect that version and push it...

Reply:
I guess we usually go via datetime anyway, so going via a bytes object shouldn't hurt.  The tuple update pattern is nice, but indeed a bit annoying unless we use it in more places.
There is a balance towards just fixing the problem at hand, and making sure the code converges to being proper.  I like to do at least a bit of the second, because chances are nobody will look at it again in years.